Carthage poa

dobertRowneySr opened the pull request:
implements #3802 proposal

### Notes
- network tests needs to be reviewed as the consensus logic has changed

mnaamani commented:
I think we have to consider actually dropping the NPoS related pallets. Since the staking pallet still is part of the runtime, the validators are still accumulating points and can be payed rewards. If this is the only problem we can just configure zero rewards as described in https://github.com/Joystream/joystream/issues/4286

dobertRowneySr commented:
> I think we have to consider actually dropping the NPoS related pallets. Since the staking pallet still is part of the runtime, the validators are still accumulating points and can be payed rewards. If this is the only problem we can just configure zero rewards as described in #4286

Implemented, see my reply on that issue about how I went about burning all the tx fees

mnaamani commented:
> > I think we have to consider actually dropping the NPoS related pallets. Since the staking pallet still is part of the runtime, the validators are still accumulating points and can be payed rewards. If this is the only problem we can just configure zero rewards as described in #4286
> 
> Implemented, see my reply on that issue about how I went about burning all the tx fees

Yep that looks correct. A neater way (instead of commenting out code) might be to have an additional type in place of 
`struct DealWithFees` say `struct BurnAllFeesAndTips`. This way cargo test will also not need to be commented for DealWithFees, just adding one more test for `BurnAllFeesAndTips` in https://github.com/Joystream/joystream/blob/carthage/runtime/src/tests/handle_fees.rs

dobertRowneySr committed:
feat: enable PoA

prevent new era from occurring, effectively:
- stopping validator set election and update
- stopping era payout from being updated
- keeping the same validator set (PoA)

mnaamani commented:
needs cargo-fmt

mnaamani commented on runtime/src/lib.rs:
```diff
@@ -466,6 +466,11 @@ impl<R: OnUnbalanced<NegativeImbalance>> OnUnbalanced<NegativeImbalance> for Dea
     }
 }
 
+pub struct DealWithFeesPoA<R>(PhantomData<R>);
```
`DealWithFeesPoA` doesn't seem to be the right name, how about `BurnAllTxFees` and lets add a test similar to the one for `DealWithFees`. Arguably that should also be renamed to reflect how it handles fees (now that we have two implementations)

mnaamani replied:
Also since for this impl we are not sending fees to anywhere it doesn't really need to be generic. So it can be simply `pub struct BurnAllTxFees {}` and then

```
    type OnChargeTransaction = CurrencyAdapter<Balances, BurnAllTxFees>;
```

mnaamani reviewed (changes requested):
I think you forgot to import `BurnAllTxFees` in the tests:

```
 error[E0433]: failed to resolve: use of undeclared type `BurnAllTxFees`
  --> runtime/src/tests/handle_fees.rs:83:9
   |
83 |         BurnAllTxFees::<Recipient>::on_unbalanceds(vec![fees, tips].into_iter());
   |         ^^^^^^^^^^^^^ not found in this scope
   |
help: consider importing this struct
   |
1  | use crate::BurnAllTxFees;
```

mnaamani commented:
clippy fix is not enough, the test is missing import of `BurnAllTxFees`

mnaamani commented on runtime/src/tests/handle_fees.rs:
```diff
@@ -47,3 +48,47 @@ fn block_author_only_receives_tips() {
+            fn on_nonzero_unbalanced(amount: NegativeImbalance) {
+                Balances::resolve_creating(&author(), amount);
+            }
+        }
```
struct Recipient and its impl can be removed here.

mnaamani commented on runtime/src/tests/handle_fees.rs:
```diff
@@ -47,3 +48,47 @@ fn block_author_only_receives_tips() {
+#[test]
+fn block_author_receives_no_tips() {
+    initial_test_ext().execute_with(|| {
+        fn author() -> AccountId {
```
you can drop this author it is not used.

dobertRowneySr replied:
It is user later to make sure that author free balances is 0

mnaamani commented on runtime/src/tests/handle_fees.rs:
```diff
@@ -47,3 +48,47 @@ fn block_author_only_receives_tips() {
 }
+
+#[test]
+fn block_author_receives_no_tips() {
```
```suggestion
fn all_tx_fees_are_burned() {
```

dobertRowneySr committed:
fix: formatting

dobertRowneySr commented:
#### How forcing new era works
- independently of the session index we are in a new era is started immediately (together with the respective election cycle)
- forcing is then set to `NotForcing` allowing the triggering of new eras only at multiple of `SessionsPerEra`

dobertRowneySr committed:
remove BurnAll the tx fees

dobertRowneySr committed:
fix: failing build

Lezek123 reviewed (changes requested):
I tested this solution, but unfortunately it looks like the era points and rewards are still accumulated by the initial validator set. This is because, contrary to our assumption, the `ActiveEra` is not actually `None` when the chain is started, but it gets initialized to era `0` regardless of the value of `Forcing`: https://github.com/paritytech/substrate/blob/master/frame/staking/src/pallet/impls.rs#L305

On the screen below you can see the initial value of `ActiveEra` (this is while still in POA):
![image](https://user-images.githubusercontent.com/19978260/193124524-0c4bb4c7-ee6f-48a8-b73b-05e8163ac986.png)

After the `staking.forceNewEra` extrinsic is called by sudo, the `ErasValidatorReward` entry for era `0` is populated based on, among others, the era length (which means the longer the initial era, the higher are the rewards accumulated):
![image](https://user-images.githubusercontent.com/19978260/193126382-39884974-3023-413e-86cd-e839b56b6f69.png)

In era `1`, which started at around block `160` a new validator set was selected, so the initial validator set was no longer producing blocks. The era length in my configuration is `60`, so the era `1` was much shorter than era `0`, but the new validator set had much higher stakes, which caused the validator rewards to increase significantly. You can notice however, that even with those increased rewards, the total reward for era `1` is lower that the reward for era `0` (because the era `0` was much longer):
![image](https://user-images.githubusercontent.com/19978260/193127563-9c74c320-6117-426b-b75f-e26da194634a.png)

And this is the `Payouts` view during era `2` (Alice and Bob were the POA validators, while the other 2 keys are the current ones):
![image](https://user-images.githubusercontent.com/19978260/193128366-342461bc-f268-4f6e-aa64-fed2b07beac2.png)

**Possible solution: Reward curve?**

I was thinking about how can this issue be addressed and while writing this summary an idea came to my mind. When the validator set was re-elected at era `1` the validators that were elected had much higher stakes than the initial authorities, which made the rewards go up, as the total stake became closer to "ideal_stake".

What if we were to set the initial authorities' stakes to `0` and construct a reward curve in such a way that when `total_stake == 0`, the inflation rate is also `0`?

This would allow us to still not need to do a runtime upgrade, because basically any validator set except the inital POA set would always have `total_stake > 0`.

I haven't tried this solution, but seems like something worth investigating.

dobertRowneySr commented:
closing in favor of https://github.com/Joystream/joystream/pull/4315
